### PR TITLE
accept replacements with branches of github forks

### LINKED ISSUE #50: Can't do module replacement with github forks
So I want to have a replacement like:

```
replace github.com/mholt/caddy-dynamicdns => github.com/francislavoie/caddy-dynamicdns caddyfile
```

So I try this:

```
xcaddy build --with github.com/mholt/caddy-dynamicdns=github.com/francislavoie/caddy-dynamicdns@caddyfile
```

But it does this:

```
...
2021/02/21 07:25:03 [INFO] exec (timeout=0s): /usr/local/go/bin/go get -d -v github.com/mholt/caddy-dynamicdns=github.com/francislavoie/caddy-dynamicdns@caddyfile
go get github.com/mholt/caddy-dynamicdns=github.com/francislavoie/caddy-dynamicdns@caddyfile: malformed module path "github.com/mholt/caddy-dynamicdns=github.com/francislavoie/caddy-dynamicdns": invalid char '='
```

This obviously isn't quite right, cause it never did the `go mod edit -replace`.

Next I tried without `@caddyfile` (my branch), and I get this:

```
2021/02/21 07:29:24 [INFO] Resolved relative replacement github.com/mholt/caddy-dynamicdns=github.com/francislavoie/caddy-dynamicdns to /root/github.com/francislavoie/caddy-dynamicdns
...
2021/02/21 07:29:24 [INFO] Replace github.com/mholt/caddy-dynamicdns => /root/github.com/francislavoie/caddy-dynamicdns
2021/02/21 07:29:24 [INFO] exec (timeout=10s): /usr/local/go/bin/go mod edit -replace github.com/mholt/caddy-dynamicdns=/root/github.com/francislavoie/caddy-dynamicdns
...
caddy imports
        github.com/mholt/caddy-dynamicdns: github.com/mholt/caddy-dynamicdns@v0.0.0-00010101000000-000000000000: replacement directory /root/github.com/francislavoie/caddy-dynamicdns does not exist
```

So `xcaddy` thinks that path is relative but it's not, it's a git repo.

I think to detect if it's relative, we should look if there's a `.` in front, so if you want to refer to something in the current dir then you do `./blah` in the replacement. That should allow github module paths to work.

The workaround obviously for now is to `git clone` the relevant repo then use a relative path on that, but it would be nice so that people can easily test branches/PRs for plugins without cloning.

### PR DESCRIPTION
**Pull request by Mohammed90**
Closes #50

### CODE REVIEW

**Comment by francislavoie** [comment]
Have you tested that this works?

```
xcaddy build --with github.com/caddyserver/forwardproxy@caddy2=github.com/klzgrad/forwardproxy@naive
```

Actual usecase from someone else who ran into this right after I reported it 😅

I'll try it out myself later when I get to my env, but impl looks about right to me.

**Comment by Mohammed90** [comment]
> Have you tested that this works?
> 
> ```
> xcaddy build --with github.com/caddyserver/forwardproxy@caddy2=github.com/klzgrad/forwardproxy@naive
> ```
> 
> Actual usecase from someone else who ran into this right after I reported it 😅
> 
> I'll try it out myself later when I get to my env, but impl looks about right to me.

Works:

```
~/projects/caddyserver/xcaddy/cmd/xcaddy ./xcaddy build --with github.com/caddyserver/forwardproxy@caddy2=github.com/klzgrad/forwardproxy@naive --
output ./caddy-fwdproxy
2021/02/26 17:18:17 [INFO] Temporary folder: /Users/mohammed/projects/caddyserver/xcaddy/cmd/xcaddy/buildenv_2021-02-26-1718.581327436
2021/02/26 17:18:17 [INFO] Writing main module: /Users/mohammed/projects/caddyserver/xcaddy/cmd/xcaddy/buildenv_2021-02-26-1718.581327436/main.go
2021/02/26 17:18:17 [INFO] Initializing Go module
2021/02/26 17:18:17 [INFO] exec (timeout=10s): /usr/local/bin/go mod init caddy
go: creating new go.mod: module caddy
go: to add module requirements and sums:
	go mod tidy
2021/02/26 17:18:17 [INFO] Replace github.com/caddyserver/forwardproxy => github.com/klzgrad/forwardproxy@naive
2021/02/26 17:18:17 [INFO] exec (timeout=10s): /usr/local/bin/go mod edit -replace github.com/caddyserver/forwardproxy=github.com/klzgrad/forwardproxy@naive
2021/02/26 17:18:17 [INFO] Pinning versions
2021/02/26 17:18:17 [INFO] exec (timeout=0s): /usr/local/bin/go get -d -v github.com/caddyserver/caddy/v2
go get: added github.com/caddyserver/caddy/v2 v2.3.0
2021/02/26 17:18:24 [INFO] Build environment ready
2021/02/26 17:18:24 [INFO] Building Caddy
2021/02/26 17:18:24 [INFO] exec (timeout=0s): /usr/local/bin/go mod tidy
go: downloading github.com/klzgrad/forwardproxy v0.0.0-20210120121422-9b4a5a242bd6
go: found github.com/caddyserver/forwardproxy in github.com/caddyserver/forwardproxy v0.0.0-00010101000000-000000000000
2021/02/26 17:18:30 [INFO] exec (timeout=0s): /usr/local/bin/go build -o /Users/mohammed/projects/caddyserver/xcaddy/cmd/xcaddy/caddy-fwdproxy -ldflags -w -s -trimpath
2021/02/26 17:19:06 [INFO] Build complete: ./caddy-fwdproxy
2021/02/26 17:19:06 [INFO] Cleaning up temporary folder: /Users/mohammed/projects/caddyserver/xcaddy/cmd/xcaddy/buildenv_2021-02-26-1718.581327436

 ~/projects/caddyserver/xcaddy/cmd/xcaddy ./caddy-fwdproxy list-modules                                                                                                               
admin.api.load
...
http.handlers.forward_proxy
```

**Review by francislavoie — APPROVED** [review]

File: cmd/xcaddy/main.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func runBuild(ctx context.Context, args []string) error {
 			})
 			if repl != "" {
 				// adjust relative replacements in current working directory since our temporary module is in a different directory
-				if !filepath.IsAbs(repl) {
+				if strings.HasPrefix(repl, ".") && !filepath.IsAbs(repl) {
```
Comment (mholt) [inline cmd/xcaddy/main.go:79]: I believe if `strings.HasPrefix(repl, ".")` is true, then `!filepath.IsAbs(repl)` is always true as well. We can probably drop the `!IsAbs()` condition.

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for this patch! Double-check my logic on my one nit; otherwise this LGTM.

File: cmd/xcaddy/main.go (hunk 1 above)
Comment (francislavoie) [reply cmd/xcaddy/main.go:79]: https://golang.org/src/path/filepath/path_unix.go?s=320:348#L2
```go
// IsAbs reports whether the path is absolute.
func IsAbs(path string) bool {
	return strings.HasPrefix(path, "/")
}
```

So yeah :+1:

Comment (Mohammed90) [reply cmd/xcaddy/main.go]: Yep, good point. Remove.

**Pull request merged**